src/console_gui.cpp

```diff
@@ -68,9 +69,52 @@ static std::deque<IConsoleLine> _iconsole_buffer;
 
 static bool TruncateBuffer();
 
+class ConsoleAutoCompletion final : public AutoCompletion {
+public:
+	using AutoCompletion::AutoCompletion;
+
+private:
+	std::vector<std::string> GetSuggestions(std::string_view prefix, std::string_view query) override
+	{
+		while (prefix.ends_with(" ")) {
+			prefix.remove_suffix(1);
+		}
+
+		std::vector<std::string> suggestions;
+
+		// We only suggest commands or aliases, so we only do it for the first token or
+		// an argument to help command.
+		if (!prefix.empty() && prefix != "help") {
+			return {};
+		}
+
+		for (const auto& [_, command] : IConsole::Commands()) {
+			if (command.name.starts_with(query)) {
+				suggestions.push_back(command.name);
+			}
+		}
+		for (const auto& [_, alias] : IConsole::Aliases()) {
+			if (alias.name.starts_with(query)) {
+				suggestions.push_back(alias.name);
+			}
+		}
+
+		return suggestions;
+	}
+
+	void ApplySuggestion(std::string_view prefix, std::string_view suggestion) override
+	{
+		if (prefix.empty()) {
+			this->textbuf->Assign(fmt::format("{} ", suggestion));
+		} else {
+			this->textbuf->Assign(fmt::format("{}{} ", prefix, suggestion));
+		}
```

Comment:
Seems a redundant distinction, as both branches do the same thing. Keep it simple, I say

Reply:
Done